src/Raven.Server/Documents/Indexes/Persistence/Corax/CoraxDocumentConverter.cs

```diff
@@ -0,0 +1,281 @@
+using Sparrow.Json;
+using Sparrow.Json.Parsing;
+using Sparrow.Server;
+using Sparrow.Threading;
+using Voron;
+using Encoding = System.Text.Encoding;
+
+namespace Raven.Server.Documents.Indexes.Persistence.Corax
+{
+    public sealed class CoraxDocumentConverter : ConverterBase
+    {
+        private readonly ByteStringContext _allocator;
+        private readonly List<IDisposable> _slicesToDispose;
+        private readonly Dictionary<Slice, int> _knownFields;
+        private static readonly byte[] _trueLiteral = Encoding.UTF8.GetBytes("true"); 
+        private static readonly byte[] _falseLiteral = Encoding.UTF8.GetBytes("false");
+        private static readonly StandardFormat _standardFormat = new StandardFormat('g');
+        private static readonly StandardFormat _timeSpanFormat = new StandardFormat('c');
+        public CoraxDocumentConverter(
+                Index index, 
+                bool indexImplicitNull = false, 
+                bool indexEmptyEntries = true, 
+                string keyFieldName = null, 
+                bool storeValue = false, 
+                string storeValueFieldName = Constants.Documents.Indexing.Fields.ReduceKeyValueFieldName) :
+            base(index, storeValue)
+        {
+            _allocator = new ByteStringContext(SharedMultipleUseFlag.None);
+            _slicesToDispose = new();
+            _knownFields = GetKnownFields();
+        }
+
+        public Dictionary<Slice, int> GetKnownFields()
+        {
+            if (_knownFields != null)
+                return _knownFields;
+            var knownFields = new Dictionary<Slice, int>();
+            foreach (var field in _fields.Values)
+            {
+                _slicesToDispose.Add(Slice.From(_allocator, field.Name, ByteStringType.Immutable, out var value));
+                knownFields.Add(value, field.Id);
+            }
+
+            return knownFields;
+        }
+        
+        public Span<byte> InsertDocumentFields(LazyStringValue key, LazyStringValue sourceDocumentId, object doc, JsonOperationContext indexContext, out LazyStringValue id)
+        {
+            var document = (Document)doc;
+            using ( _allocator.Allocate(document.Data.Size * 2, out ByteString buffer))
+            {
+                var entryWriter = new IndexEntryWriter(buffer.ToSpan(), _knownFields);
+                id = document.LowerId;
+                bool shouldSkip;
+                
+                
+                //TODO maciej - please look at this.
+                // this is reference to list for EnumerableWritingScope due to making it persistence during indexing document. 
+                // We want avoid allocation for every enumerable in doc, but if think it should be persistence during whole indexing process.
+                // Where should we put this and when release it?
+                List<int> stringsLength = new List<int>(128);
+                var scope = new SingleEntryWriterScope(stringsLength, _allocator);
```

Comment:
Given the above comment about the reuse of the list - I wonder if the `SingleEntryWriterScope` shoudn't be disposable and the dispose will be called after we write entries to the index. Then we could release the list. So maybe the scope should be just passed as argument here but wrapped into `using` outside

Reply:
Temporarly.